From 0db2d3dfd736030588d8ca844faf54c599a6401e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:08:18 -0400
Subject: [usb] sierra: fix characters being dropped at close

Message-id: <1410887463-20674-427-git-send-email-dzickus@redhat.com>
Patchwork-id: 95571
O-Subject: [RHEL7 PATCH 426/591] USB: sierra: fix characters being dropped at close
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 93670599fc52217dcf8a69832faf66171cdb9581
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:22:55 2014 +0200

    USB: sierra: fix characters being dropped at close

    Fix characters potentially being dropped at close due to missing
    chars_in_buffer implementation.

    Note that currently the write urbs are not even killed at close (will be
    fixed separately), but this could still lead to dropped data since we
    have lowered DTR/RTS.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/sierra.c b/drivers/usb/serial/sierra.c
index 3748034..ed43b18 100644
--- a/drivers/usb/serial/sierra.c
+++ b/drivers/usb/serial/sierra.c
@@ -675,6 +675,23 @@ static int sierra_write_room(struct tty_struct *tty)
  return 2048;
 }
 
+static int sierra_chars_in_buffer(struct tty_struct *tty)
+{
+ struct usb_serial_port *port = tty->driver_data;
+ struct sierra_port_private *portdata = usb_get_serial_port_data(port);
+ unsigned long flags;
+ int chars;
+
+ /* NOTE: This overcounts somewhat. */
+ spin_lock_irqsave(&portdata->lock, flags);
+ chars = portdata->outstanding_urbs * MAX_TRANSFER;
+ spin_unlock_irqrestore(&portdata->lock, flags);
+
+ dev_dbg(&port->dev, "%s - %d\n", __func__, chars);
+
+ return chars;
+}
+
 static void sierra_stop_rx_urbs(struct usb_serial_port *port)
 {
  int i;
@@ -1060,6 +1077,7 @@ static struct usb_serial_driver sierra_device = {
  .dtr_rts    = sierra_dtr_rts,
  .write             = sierra_write,
  .write_room        = sierra_write_room,
+ .chars_in_buffer   = sierra_chars_in_buffer,
  .set_termios       = sierra_set_termios,
  .tiocmget          = sierra_tiocmget,
  .tiocmset          = sierra_tiocmset,
-- 
1.7.1